complex.go

```diff
@@ -374,7 +394,7 @@ func (s *ObjectShape) validateProperties(ctxPath string, props map[string]interf
 		if found {
 			continue
 		}
-		if !found && restrictedAdditionalProperties {
+		if restrictedAdditionalProperties {
```

Comment:
I think it might be better to actually move this condition back after `s.validatePatternProperty` and put the comment like "Will never be true if pattern properties are present.". Because otherwise this looks like a logic gap and I overthought this logic during refactoring

Reply:
Will be continued in the next PRs